File: red/src/key-request.ts
import {lstatSync,openSync,readSync,closeSync,constants} from 'node:fs';
import {homedir} from 'node:os';
import {join,extname} from 'node:path';
import {createHash} from 'node:crypto';
import providers from '../resources/providers.json';
import recipes from '../resources/provider-recipes.json';
type Map=Record<string,any>;
const missing=(v:any)=>v==null||(typeof v==='string'&&(!v.trim()||v.trim().toUpperCase()==='REPLACE_ME'));
function publicKey(value:any):string {
 const error=()=>{throw Error('invalid external SSH public key');};
 if(typeof value!=='string')return error();value=value.trim();const parts=value.split(/\s+/);
 if(parts.length<2||/[\r\n]/.test(value)||!['ssh-ed25519','ssh-rsa','ecdsa-sha2-nistp256','ecdsa-sha2-nistp384','ecdsa-sha2-nistp521'].includes(parts[0]))return error();
 if(!/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(parts[1]))return error();
 const blob=Buffer.from(parts[1],'base64');if(blob.length<4)return error();const n=blob.readUInt32BE(0);
 if(blob.length<=4+n||!blob.subarray(4,4+n).equals(Buffer.from(parts[0])))return error();return value;
}
export function key_request(opts:Map,prepared:Map,environment:Record<string,string|undefined>|null=null) {
 if(prepared.mode==='managed')return {mode:'managed',public_key:prepared.public_key};
 if(prepared.mode!=='external')throw Error('invalid compute key request');
 const provider=opts['provider-compute'];if(typeof provider!=='string'||!Object.hasOwn(recipes,provider))throw Error('compute provider recipe unavailable');
 const kind=(providers.compute as Map)[provider]['ssh-aliases']?.[prepared.setting]??(recipes as Map)[provider].external_key_kind;
 const refs=Array.isArray(prepared.reference)?prepared.reference:[prepared.reference];
 if(kind==='ids') {
  if(!refs.length||refs.some(v=>!(typeof v==='string'&&!missing(v)||typeof v==='number'&&Number.isSafeInteger(v)&&v>0)))throw Error('invalid external SSH key reference');
  return {mode:'external',ids:structuredClone(refs),reference:refs[0]};
 }
 if(refs.length!==1||typeof refs[0]!=='string'||missing(refs[0]))throw Error('one external SSH public key is required');
 if(kind==='fingerprint_file'&&(!refs[0].endsWith('.pub')||/[\0\r\n]/.test(refs[0])))throw Error('external SSH key must name a regular .pub file');
 if(opts['red/event']==='build'||opts['red/dry-run']===true){if(kind==='fingerprint_file'){const value=Array(16).fill('00').join(':');return {mode:'external',ids:[value],reference:value};}return {mode:'external',public_key:'ssh-ed25519 PLACEHOLDER managed-by-colors'};}
 let value:string;
 if(kind==='content')value=publicKey(refs[0]);
 else if(kind==='public_file'||kind==='fingerprint_file') {
  const env=environment??process.env;const path=refs[0].startsWith('~/')?join(env.HOME||homedir(),refs[0].slice(2)):refs[0];
  let valid=false;try{valid=extname(path)==='.pub'&&lstatSync(path).isFile();}catch{}
  if(!valid)throw Error('external SSH key must name a regular .pub file');
  let fd:number|undefined;
  try{fd=openSync(path,constants.O_RDONLY|constants.O_NOFOLLOW);const bytes=Buffer.alloc(262149);let count=0,n=0;while(count<bytes.length&&(n=readSync(fd,bytes,count,bytes.length-count,null))>0)count+=n;
   const content=new TextDecoder('utf-8',{fatal:true,ignoreBOM:true}).decode(bytes.subarray(0,count));if([...content].length>65536)throw Error();value=publicKey(content);
  }catch{throw Error('invalid external SSH public key file');}finally{if(fd!==undefined)closeSync(fd);}
 }else throw Error('unsupported external SSH key reference');
 if(kind==='fingerprint_file'){const fingerprint=createHash('md5').update(Buffer.from(value.split(/\s+/)[1],'base64')).digest('hex').match(/../g)!.join(':');return {mode:'external',ids:[fingerprint],reference:fingerprint};}
 return {mode:'external',public_key:value};
}

File: blue/src/colors_compute/key_request.py
"""Resolve explicitly selected public key references for provider requests."""
import base64
import hashlib
from ._copy import deepcopy
from importlib.resources import files
import json
import os
from pathlib import Path
import struct

from .contract import _missing, registry
from .ssh import PLACEHOLDER


def _public(value):
    if not isinstance(value, str):
        raise ValueError('invalid external SSH public key')
    value = value.strip()
    parts = value.split()
    if len(parts) < 2 or '\n' in value or '\r' in value or parts[0] not in ('ssh-ed25519', 'ssh-rsa', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384', 'ecdsa-sha2-nistp521'):
        raise ValueError('invalid external SSH public key')
    try:
        blob = base64.b64decode(parts[1], validate=True)
        length = struct.unpack('!I', blob[:4])[0]
        if blob[4:4+length] != parts[0].encode() or len(blob) <= 4 + length:
            raise ValueError()
    except (ValueError, struct.error):
        raise ValueError('invalid external SSH public key') from None
    return value


def key_request(opts, prepared, environment=None):
    """No private file is opened. Planning never opens a public file either."""
    if prepared.get('mode') == 'managed':
        return {'mode': 'managed', 'public_key': prepared['public_key']}
    if prepared.get('mode') != 'external':
        raise ValueError('invalid compute key request')
    recipes = json.loads(files('colors_compute').joinpath('provider-recipes.json').read_text())
    provider = opts.get('provider-compute')
    if not isinstance(provider, str) or provider not in recipes:
        raise ValueError('compute provider recipe unavailable')
    kind = registry()['compute'][provider].get('ssh-aliases', {}).get(prepared.get('setting'), recipes[provider]['external_key_kind'])
    reference = prepared.get('reference')
    references = reference if isinstance(reference, list) else [reference]
    if kind == 'ids':
        if not references or any(not (isinstance(item, str) and not _missing(item) or type(item) in (int, float) and 0 < item <= 9007199254740991 and int(item) == item) for item in references):
            raise ValueError('invalid external SSH key reference')
        return {'mode': 'external', 'ids': deepcopy(references), 'reference': references[0]}
    if len(references) != 1 or not isinstance(references[0], str) or _missing(references[0]):
        raise ValueError('one external SSH public key is required')
    if kind == 'fingerprint_file' and (not references[0].endswith('.pub') or any(c in references[0] for c in '\x00\r\n')):
        raise ValueError('external SSH key must name a regular .pub file')
    if opts.get('blue/event') == 'build' or opts.get('blue/dry-run') is True:
        if kind == 'fingerprint_file':
            value = ':'.join(['00'] * 16)
            return {'mode': 'external', 'ids': [value], 'reference': value}
        return {'mode': 'external', 'public_key': PLACEHOLDER}
    if kind == 'content':
        public = _public(references[0])
    elif kind in ('public_file', 'fingerprint_file'):
        name = references[0]
        env = os.environ if environment is None else environment
        if name.startswith('~/'):
            name = os.path.join(env.get('HOME') or str(Path.home()), name[2:])
        path = Path(name)
        if path.suffix != '.pub' or path.is_symlink() or not path.is_file():
            raise ValueError('external SSH key must name a regular .pub file')
        try:
            with os.fdopen(os.open(path, os.O_RDONLY | os.O_NOFOLLOW), 'r', encoding='utf-8') as stream:
                content = stream.read(65537)
            if len(content) > 65536:
                raise ValueError()
            public = _public(content)
        except (OSError, UnicodeError, ValueError):
            raise ValueError('invalid external SSH public key file') from None
    else:
        raise ValueError('unsupported external SSH key reference')
    if kind == 'fingerprint_file':
        value = ':'.join(f'{byte:02x}' for byte in hashlib.md5(base64.b64decode(public.split()[1])).digest())
        return {'mode': 'external', 'ids': [value], 'reference': value}
    return {'mode': 'external', 'public_key': public}
